Fixes issue/3846-auth-error-changes

anitaa1990 opened the pull request:
Fixes #3846. This PR adds the option to fetch the user role for a store and displays an error screen if they do not have admin or shop manager access. 

### Changes
- Adds a separate class `UserEligibilityFetcher` (forgive the name, please!) that fetches the user info from FluxC.
- This class is called: 
  - Every time the app is completely restarted 
  - When the user logs in for the first time. 
- If the API request is successful, we store a Boolean flag and the user email to `AppPrefs`. 
- Every time the app is opened, we check this Boolean flag to determine if the user has sufficient access.
- An error screen is displayed if they do not have admin or shop manager access.

### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/119083534-c234f300-ba1d-11eb-868b-8f8dbb58a9d8.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/119083549-ca8d2e00-ba1d-11eb-8bea-bac167abf7c5.png" width="300"/>


### Testing
#### Perquisites
- At the moment we only support Admin and Shop Manager role. Create multiple user accounts with different user roles on your test site. 
- You can install the [User Role Editor](https://user-images.githubusercontent.com/22608780/116804544-18a8c300-ab3d-11eb-880e-a60381940a82.png) plugin on your test site. This allows you to create multiple custom roles and assign multiple roles to the same user.

#### Scenario 1
- Login to your test site using an admin or shop manager credentials.
- You should be able to use the app as is without any issues.

#### Scenario 2
- Login to your test site using any credentials that does not include admin or shop manager.
- You should be able to see an error screen when you open the app.
- Click on the back button and verify that you are exited from the app.
- Click on the Help button and verify that you are able to see the support screen.
- Click on the `Retry` button and verify that, if the account has insufficient access, you see an error Snackbar. Try changing the user's access from `wp-admin` and click on the Retry button again. You should be redirected to the Stats screen.
- Click on the `Login to another account` button and verify that you are redirected to the Prologue screen.

#### Unit tests
- Run `UserEligibilityErrorViewModelTest` class.
- Run `UserEligibilityFetcherTest` class.

@designsimply I added you as a reviewer since you were the one that caught the bug in the first place 😄 

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Updated FluxC hash

anitaa1990 committed:
Added separate class to fetch user info

anitaa1990 committed:
Added methods to display error screen if user is not eligible

anitaa1990 committed:
Set default prefs value to true i.e. user is always eligible until API data is fetched

anitaa1990 committed:
Added logic to fetch user info when the app is completely restarted

anitaa1990 committed:
Added logic to display error screen when user does not have the right access

anitaa1990 committed:
Added user email to app prefs

anitaa1990 committed:
Added unit tests for UserEligibilityFetcherTest

anitaa1990 committed:
Added viewModel + unit test for fetching user info from db

anitaa1990 committed:
Added UI for user eligiblity error screen

anitaa1990 committed:
Added click listener for help button

anitaa1990 committed:
Added logic to handle back button click

anitaa1990 committed:
Moved the get user info from db method to UserEligibilityFetcher

anitaa1990 committed:
Added click listener the the retry button

anitaa1990 committed:
Added click listener for learn more button

anitaa1990 committed:
Added click listener for logout button

anitaa1990 committed:
Updated unit tests

anitaa1990 committed:
Updated illustrations for eligiblity error screen

anitaa1990 committed:
Added logic to fetch user info after user is logged in

adamzelinski reviewed (changes requested):
This is looking really good thanks @anitaa1990! I did notice that the background on the light version needs the white background please and also it seems the headers and dark buttons are a little different from the designs. 

<img width="1815" alt="Screen Shot 2021-05-21 at 3 17 03 pm" src="https://user-images.githubusercontent.com/74938219/119085837-d261c800-ba47-11eb-989a-99783dc4df91.png">

anitaa1990 commented:
Thank you for the quick review @adamzelinski ❤️ 

>  I did notice that the background on the light version needs the white background please and also it seems the headers and dark buttons are a little different from the designs.

I missed the light version background. Updated that and this is how it looks now. 
<img src="https://user-images.githubusercontent.com/22608780/119089935-90755980-ba28-11eb-890f-e6bb6b68cf63.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/119089978-9f5c0c00-ba28-11eb-864b-abde4c8e1579.png" width="300"/>


I didn't add a title to the screen since this is something we check after the user is logged in, so `Log in` as a header here feels misleading. I couldn't think of anything else appropriate so just left it blank 😅 

Regarding the text styles, so the bottom buttons (`Retry` and `Login with another account`) are using the same styles as the ones defined for all the login screens. We're just reusing the same layout here to be consistent across the different error screens. Do we want to change it for all screens? We can do that but it out of scope for this PR. I'm happy to tackle the change in a different PR, if that's alright with you?

anitaa1990 committed:
Fixed design changes requested

anitaa1990 committed:
Added release notes

ghost commented:
No that's cool @anitaa1990 thanks for updating the bg style the rest is fine to use the same styles

nbradbury commented:
If I go to settings and switch to a store I don't have the correct user role for, I see the toast stating that the store was switched.

![toast](https://user-images.githubusercontent.com/3903757/119343782-e8000900-bc64-11eb-830a-2e1f8b4b2f53.png)

When I tap the back button from there, I then see the screen stating I don't have the correct user role. Shouldn't the user role verification happen before then?

And if I tap the back button on that screen, it closes the app. The app also closes if I tap the back button on that screen during the initial login flow. In that situation, it should probably return the user to the first login screen.

nbradbury commented:
This seems unnecessarily wordy. How about: "This app supports only Administrator and Shop Manager user roles. Please contact your store owner to upgrade your role."

![role](https://user-images.githubusercontent.com/3903757/119344647-e7b43d80-bc65-11eb-8867-c89131707c6c.png)

Also, the toast that appears when you tap retry says "user access" - how about "user role" instead for consistency?

nbradbury commented:
When the device is in landscape I see this:

![landscape](https://user-images.githubusercontent.com/3903757/119345139-7b860980-bc66-11eb-803d-4728c8f36116.png)

In similar screens elsewhere in the app we hide the image in landscape (but not on tablets). Should we do that here? @adamzelinski

nbradbury commented:
I'm not sure how, but I was able to login to the app with a "Contributor" user role. The dashboard showed a "Error fetching data" toast and the log showed "Error fetching visitor stats: user cannot view stats."

![Screenshot_20210524_081020](https://user-images.githubusercontent.com/3903757/119346054-b3418100-bc67-11eb-8a18-c29420b42010.png)

I then went to settings, selected the same store, and tapped Done. When I returned to the dashboard, the app crashed with this:

```
 Process: com.woocommerce.android, PID: 2882
    java.lang.NullPointerException
        at com.woocommerce.android.ui.mystore.MyStoreFragment.getTabLayout(MyStoreFragment.kt:83)
        at com.woocommerce.android.ui.mystore.MyStoreFragment.getActiveGranularity(MyStoreFragment.kt:76)
        at com.woocommerce.android.ui.mystore.MyStoreFragment.access$getActiveGranularity$p(MyStoreFragment.kt:46)
        at com.woocommerce.android.ui.mystore.MyStoreFragment$refreshMyStoreStats$$inlined$run$lambda$1.invokeSuspend(MyStoreFragment.kt:319)
```

(I'm guessing that crash is unrelated to this PR.)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/UserEligibilityErrorFragment.kt:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    override fun onRequestAllowBackPress(): Boolean {
+        activity?.finish()
```
I mentioned this already, but here it probably makes more sense to return to the previous screen instead of closing the app when the back button is pressed.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/UserEligibilityErrorFragment.kt:
```diff
@@ -0,0 +1,156 @@
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        if (item.itemId == R.id.help) {
+            startActivity(HelpActivity.createIntent(requireActivity(), Origin.USER_ELIGIBILITY_ERROR, null))
```
WDYT about logging the user role here? That could help HEs if the user contacts us from the help activity.

anitaa1990 replied:
That's a very good idea! I added that in e044cc0

nbradbury commented:
@anitaa1990 Done with my first round!

ghost commented:
> "This app supports only Administrator and Shop Manager user roles. Please contact your store owner to upgrade your role."

> Also, the toast that appears when you tap retry says "user access" - how about "user role" instead for consistency?

I agree @nbradbury could we please make both of these changes.

ghost commented:
> In similar screens elsewhere in the app we hide the image in landscape (but not on tablets). Should we do that here? 

Yes, thanks @nbradbury we will hide the image in landscape (but not on tablets)

anitaa1990 committed:
Updated error message text and added screen for landscape mode

anitaa1990 committed:
Added logic to fetch user role after site verification

anitaa1990 committed:
Added user role when creating support ticket

anitaa1990 commented:
Thank you for the review @nbradbury!

> In similar screens elsewhere in the app we hide the image in landscape (but not on tablets). Should we do that here?

Good point! I fixed this in 35b2981 

> This seems unnecessarily wordy. How about: "This app supports only Administrator and Shop Manager user roles. Please contact your store owner to upgrade your role."

👍 Done in 35b2981

> I'm not sure how, but I was able to login to the app with a "Contributor" user role. The dashboard showed a "Error fetching data" toast and the log showed "Error fetching visitor stats: user cannot view stats."

So I'm not really able to reproduce this scenario. But I had added logic in `SitePickerActivity` to fetch user role and store it in `AppPrefs` in the background. So it might be that `MainActivity` is called before the user could be fetched and stored in the background. So I updated the logic in `SitPickerActivity` to fetch the user role synchronously. So only after the user role is fetched and stored, we would redirect to the main activity (even if it results in an error). This is done in 102a761 and that should ensure that we don't allow users without admin or shop manager role into the app after login.

> If I go to settings and switch to a store I don't have the correct user role for, I see the toast stating that the store was switched.

I thought of this one but when testing I noticed that if I don't have access to that store, it is not listed in the store switcher list. But if this is happening when switching stores, the logic added in 102a761 should help fix this since that check happens when we switch store as well 👍 

> it probably makes more sense to return to the previous screen instead of closing the app when the back button is pressed.

So we are only able to get the user role from the API after the user is logged in successfully. So by the time they see this screen, they are logged in. So clicking the back button should ideally close the app. If we redirect to the login prologue screen we would have to log them out of the app first and they would have to login again each time they open the app. If we redirect them to the `MyStore` tab, none of the APIs would work in the app so they would just keep seeing an error, similar to the experience we have now.

Ready for another round @nbradbury 👍

nbradbury commented:
~~I'm still able to login without the correct user role. In this screenshot, both of these stores are available even though that account has a Contributor role for both of them. When I tap Done, the app shows the dashboard screen with an "Error fetching data" toast.~~

nbradbury commented:
@anitaa1990 Ignore that last message, I tested the app before I pulled changes 🤦

nbradbury commented:
The changes here include a landscape layout of the eligibility screen that has no image, which means the image is missing even on landscape tablets. We wanted to still show the image on landscape tablets, since there's plenty of room for it.

![tablet](https://user-images.githubusercontent.com/3903757/119662933-e880d700-bdff-11eb-81ec-cadd12f85fea.png)

Rather than using a separate layout, WDYT about sticking with a single layout and programmatically hiding the image when the orientation is landscape and the device isn't a tablet?

nbradbury commented:
@anitaa1990 I closed the app when the eligibility screen was showing for a user that wasn't eligible for any stores they're a member of, and when I restarted it crashed with this:

```
java.lang.IllegalStateException: SelectedSite.get() was accessed before being initialized - siteId -1.
    Consider calling selectedSite.exists() to ensure site exists prior to calling selectedSite.get().
        at com.woocommerce.android.tools.SelectedSite.get(SelectedSite.kt:36)
        at com.woocommerce.android.ui.common.UserEligibilityFetcher.fetchUserInfo(UserEligibilityFetcher.kt:34)
        at com.woocommerce.android.ui.common.UserEligibilityFetcher$fetchUserEligibility$1.invokeSuspend(UserEligibilityFetcher.kt:27)
```

I had to clear storage in order to open the app again.

nbradbury commented:
@anitaa1990 I just noticed the user name and role aren't appearing on the eligibility screen.

![Screenshot_20210526_111920](https://user-images.githubusercontent.com/3903757/119686576-75359000-be14-11eb-9406-4204bab5572f.png)

On the previous screen, though, they appear just fine.

![Screenshot_20210526_111504](https://user-images.githubusercontent.com/3903757/119686386-491a0f00-be14-11eb-956b-be8487b5097c.png)

anitaa1990 committed:
Added logic to display imaage only if orientation is landscapr and not a tablet

anitaa1990 committed:
fix merge conflict from develop

anitaa1990 commented:
Hey @nbradbury! Thanks for the review! 

> The changes here include a landscape layout of the eligibility screen that has no image, which means the image is missing even on landscape tablets. We wanted to still show the image on landscape tablets, since there's plenty of room for it.

I like the idea of having a standard way to display images in tablets and landscape mode. I only made the changes at the moment to this screen but long term we might want to find a standard way to achieve this. Thanks for the suggestion! Fixed in abf5c0d.

> @anitaa1990 I closed the app when the eligibility screen was showing for a user that wasn't eligible for any stores they're a member of, and when I restarted it crashed with this:

I'm not able to reproduce this. There seems to be a lot of issues that keep happening when I try to fix the merge conflicts in this PR. I finally had to force push to fix those changes in e70f410. Can you test again and see if this is happening please? 

> @anitaa1990 I just noticed the user name and role aren't appearing on the eligibility screen.

I noticed that too! It should be fixed in abf5c0d. I also noticed that the app was crashing when clicking on the Help button and that was fixed too!

Ready for another round 👍

nbradbury commented:
@anitaa1990 I can't reproduce that crash now, but I'm still not seeing the user info. The problem is [here in the viewmodel](https://github.com/woocommerce/woocommerce-android/pull/4074/files#diff-1c3ea75ddd6329dc433d5b6eea31160ff412b1f3a43d589d9b2a819d31e202bdR52), because the user email is empty.

<img width="538" alt="Screen Shot 2021-05-26 at 1 00 01 PM" src="https://user-images.githubusercontent.com/3903757/119701517-6a81f780-be22-11eb-9835-fd8384d4282e.png">

**Update:** I just noticed that if you view the help screen then hit the back button, the user name shows up.

![Screenshot_1622049354](https://user-images.githubusercontent.com/3903757/119703534-9f8f4980-be24-11eb-8810-403222bdd960.png)

anitaa1990 committed:
Merge branch 'develop' into issue/3846-auth-error-changes-take2

anitaa1990 commented:
> @anitaa1990 I can't reproduce that crash now, but I'm still not seeing the user info. The problem is here in the viewmodel, because the user email is empty.


~Hey @nbradbury I'm not actually able to reproduce that. Would you be able to add my email (reallychumma1@gmail.com) to your test site so I can check it out? As such, if we don't really have the email stored, we won't be able to get the user info so the user name and role would not be displayed. But I'm not sure why the email is empty and even more weird, why it's not showing the role and why it's showing after coming back from the Help screen.~

Scratch ☝️ I was able to reproduce the same issue with just the `Contributor` role. The email was not getting saved after the user logs in for the first time so it was blank, which was causing the username to be blank. Fixed in a1ea303.

Ready for another round 😅

anitaa1990 committed:
Fixed blank user name issue

anitaa1990 committed:
Merge branch 'issue/3846-auth-error-changes-take2' of https://github.com/woocommerce/woocommerce-android into issue/3846-auth-error-changes-take2

anitaa1990 committed:
Updated unit test

anitaa1990 committed:
Merge branch 'develop' into issue/3846-auth-error-changes-take2

nbradbury reviewed (approved):
@anitaa1990 I think you'll be delighted to hear that this all looks good now! :shipit:
